processing/src/main/java/io/druid/query/lookup/LookupReferencesManager.java

```diff
@@ -52,333 +56,313 @@
-        LOGGER.debug("Removed lookup [%s]", lookupName);
-        if (lookupSnapshotTaker != null) {
-          lookupSnapshotTaker.takeSnapshot(getAllAsList());
-        }
-        return lookupExtractorFactory.close();
-      }
-    }
-    finally {
-      startStopLock.readLock().unlock();
+    if (lookupSnapshotTaker != null) {
+      List<LookupBean> lookups = Lists.newArrayList(
+          Collections2.transform(
+              lookupMap.entrySet(),
+              new Function<Map.Entry<String, LookupExtractorFactoryContainer>, LookupBean>()
+              {
+                @Nullable
+                @Override
+                public LookupBean apply(
+                    @Nullable
+                        Map.Entry<String, LookupExtractorFactoryContainer> input
+                )
+                {
+                  final LookupBean lookupBean = new LookupBean(input.getKey(), null, input.getValue());
+                  return lookupBean;
+                }
+              }
+          )
+      );
+
+      lookupSnapshotTaker.takeSnapshot(lookups);
     }
-    return false;
   }
 
-  /**
-   * @param lookupName key to fetch the reference of the object {@link LookupExtractor}
-   *
-   * @return reference of {@link LookupExtractorFactory} that correspond the {@code lookupName} or null if absent
-   *
-   * @throws IllegalStateException if the {@link LookupReferencesManager} is closed or did not start yet
-   */
-  @Nullable
-  public LookupExtractorFactory get(String lookupName)
+  @VisibleForTesting
+  interface Notice
   {
-    try {
-      startStopLock.readLock().lockInterruptibly();
-    }
-    catch (InterruptedException e) {
-      throw Throwables.propagate(e);
-    }
-    try {
-      final LookupExtractorFactory lookupExtractorFactory = lookupMap.get(lookupName);
-      assertStarted();
-      return lookupExtractorFactory;
-    }
-    finally {
-      startStopLock.readLock().unlock();
-    }
+    void handle();
   }
 
-  /**
-   * @return Returns {@link Map} containing a copy of the current state.
-   *
-   * @throws ISE if the is is closed or did not start yet.
-   */
-  public Map<String, LookupExtractorFactory> getAll()
+  private class LoadNotice implements Notice
   {
-    try {
-      startStopLock.readLock().lockInterruptibly();
-    }
-    catch (InterruptedException e) {
-      throw Throwables.propagate(e);
-    }
-    try {
-      assertStarted();
-      return Maps.newHashMap(lookupMap);
+    String lookupName;
+    LookupExtractorFactoryContainer lookupExtractorFactoryContainer;
+
+    public LoadNotice(String lookupName, LookupExtractorFactoryContainer lookupExtractorFactoryContainer)
+    {
+      this.lookupName = lookupName;
+      this.lookupExtractorFactoryContainer = lookupExtractorFactoryContainer;
     }
-    finally {
-      startStopLock.readLock().unlock();
+
+    @Override
+    public void handle()
+    {
+      Preconditions.checkState(lifecycleLock.isStarted());
```

Comment:
This check is not needed because one is already done in the main loop in mainThread.